keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1543 @@
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+severity, levels, failure policies, type safety, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  constraints:
+    rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+    - expression: "object.spec.replicas <= config.maxReplicas"
+      # ...other rule related fields here...
+```
+
+The `spec.config` field of the `ValidatingAdmissionPolicy` references the
+"configuration CRD" for this admission policy. For this example, the
+configuration CRD is `ReplicaLimit`.
+
+Note: This is a "Bring Your Own CRD" design. The admission policy definition
+author is responsible for providing the `ReplicaLimit` configuration CRD.
+
+`spec.kind` specifies that the CEL expressions in `validations.expression` have
+access to the fields of the `v1.Deployment` kind.
+
+To configure an admission policy, "policy configurations" of the configuration
+CRD kind are created.
+
+For example:
+
+```yaml
+# Policy configurations
+apiVersion: rules.example.com/v1
+kind: ReplicaLimit
+metadata:
+  name: "replica-limit-test.example.com"
+spec:
+  match:
+    namespaceSelectors:
+    - key: environment,
+      operator: In,
+      values: ["test"]
+  maxReplicas: 3
+```
+
+This `replica-limit-test.example.com` policy configuration limits deployments to
+a max of 3 repliacas in all namespaces in the test environment.
+
+An admission policy may have multiple configurations. In this example an
+additional `ReplicaLimit` could be created to configure the "prod" environment
+to have a maxReplicas limit of 100.
+
+This design separates admission policy _definition_ from _configuration_. This
```

Comment:
Is the _configuration_ cluster scoped or namespace scoped? 
Might be good to add a snippet about how we plan to do versioning for the policy _definition_.